File: Java/src/pomocno/MjesanjeAdapter.java
package pomocno;

import com.google.gson.*;
import model.Mjesanje;
import model.Partija;

import java.lang.reflect.Type;

public class MjesanjeAdapter implements JsonSerializer<Mjesanje>, JsonDeserializer<Mjesanje> {
    @Override
    public JsonElement serialize(Mjesanje src, Type typeOfSrc, JsonSerializationContext context) {
       return null;
    }

    @Override
    public Mjesanje deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
            throws JsonParseException {
        JsonObject jsonObject = json.getAsJsonObject();
        try {
        if (jsonObject.has("bodovaTreciUnos")){
            return context.deserialize(json, Class.forName("model.MjesanjeTriUnosa"));
        }else{
            return context.deserialize(json, Class.forName("model.MjesanjeDvaUnosa"));
        }
        } catch (ClassNotFoundException cnfe) {
            throw new JsonParseException("Unknown element type: ", cnfe);
        }
    }
}

File: Kotlin/src/main/kotlin/pomocno/MjesanjeAdapter.kt
package pomocno

import com.google.gson.*
import model.Mjesanje
import java.lang.reflect.Type

/**
 * @author Matija Skrletović
 **/

class MjesanjeAdapter : JsonSerializer<Mjesanje?>, JsonDeserializer<Mjesanje> {
    override fun serialize(
        src: Mjesanje?,
        typeOfSrc: Type,
        context: JsonSerializationContext
    ): JsonElement? {
        return null
    }

    @Throws(JsonParseException::class)
    override fun deserialize(
        json: JsonElement,
        typeOfT: Type,
        context: JsonDeserializationContext
    ): Mjesanje {
        val jsonObject = json.asJsonObject
        return try {
            if (jsonObject.has("bodovaTreciUnos")) {
                context.deserialize(json, Class.forName("MjesanjeTriUnosa"))
            } else {
                context.deserialize(json, Class.forName("MjesanjeDvaUnosa"))
            }
        } catch (cnfe: ClassNotFoundException) {
            throw JsonParseException("Unknown element type: ", cnfe)
        }
    }
}